feat(schema): PDE-5719 split FieldSchema into InputFieldSchema, OutputFieldSchema, and AuthFieldSchema (#957)
zapier/zapier-platform: packages/schema/docs/build/schema.md

-----

## /AuthFieldSchema

Field schema specialized for authentication fields.

#### Details

* **Type** - `object`
* [**Source Code**](https://github.com/zapier/zapier-platform/blob/zapier-platform-schema@16.3.0/packages/schema/lib/schemas/AuthFieldSchema.js)

#### Properties

Key | Required | Type | Description
--- | -------- | ---- | -----------
`key` | **yes** | `string` | A unique machine readable key for this value (IE: "fname").

> `isNotSecret` | no | `boolean` | Indicates if this authentication field is safe (not secret).

Reviewer: Nitpick, but perhaps drop `t` and give example of what it is safe for:

```suggestion
`isNoSecret` | no | `boolean` | Indicates if this authentication field is safe to e.g. be stored without encryption or displayed (not a secret).
```

`label` | no | `string` | A human readable label for this value (IE: "First Name").
`helpText` | no | `string` | A human readable description of this value (IE: "The first part of a full name."). You can use Markdown.

> `type` | no | `string` in (`'string'`, `'number'`, `'boolean'`, `'datetime'`, `'password'`, `'copy'`) | The type of this value. Field type of `file` will accept either a file object or a string. If a URL is provided in the string, Zapier will automatically make a GET for that file. Otherwise, a .txt file will be generated.

Reviewer: The description refers to `file` even though that is no listed as a valid value:

```suggestion
`type` | no | `string` in (`'string'`, `'number'`, `'boolean'`, `'datetime'`, `'password'`, `'copy'`) | The type of this value.
```

> `required` | no | `boolean` | If this value is required or not. This defaults to True.

Reviewer: ```suggestion
`required` | no | `boolean` | If this value is required or not. This defaults to `true`.
```

> `placeholder` | no | `string` | An example value that is not saved.

Reviewer: I do believe we also support `default`, which is no longer listed.

-----

## /AuthFieldSchema

Field schema specialized for authentication fields.

#### Details

* **Type** - `object`
* [**Source Code**](https://github.com/zapier/zapier-platform/blob/zapier-platform-schema@16.2.0/packages/schema/lib/schemas/AuthFieldSchema.js)

#### Properties

Key | Required | Type | Description
--- | -------- | ---- | -----------
`key` | **yes** | `string` | A unique machine readable key for this value (IE: "fname").
`label` | no | `string` | A human readable label for this value (IE: "First Name").
`helpText` | no | `string` | A human readable description of this value (IE: "The first part of a full name."). You can use Markdown.

> `type` | no | `string` in (`'string'`, `'text'`, `'integer'`, `'number'`, `'boolean'`, `'datetime'`, `'file'`, `'password'`, `'copy'`, `'code'`) | The type of this value. Use `string` for basic text input, `text` for a large, `<textarea>` style box, and `code` for a `<textarea>` with a fixed-width font. Field type of `file` will accept either a file object or a string. If a URL is provided in the string, Zapier will automatically make a GET for that file. Otherwise, a .txt file will be generated.

Reviewer: This should be limited to what actually works.

Just from [rendering](https://cdn.zappy.app/07033c49fc406de8261e71fbf8108335.gif) it seems `code` doesn't, but I also doubt `file` would actually process well.

Author: That's really neat! How did you get it to render? I tried running the monolith and editor locally then create an app using CLI locally, but I wasn't able to see it via the editor, any tips would be awesome!

Reviewer: Yes, that's what I did as well.

Reviewer 2: Argh, Natay just called my attention to this suggestion being for AuthFieldSchema and not necessarily for InputFieldSchema. I agree then. Thanks @FokkeZB

Author: Thanks both for the guidance! I was able to [render](https://cdn.zappy.app/dab1dd0f286c5e998a7dd650aa6ccfad.gif) and keep only what works.

> `required` | no | `boolean` | If this value is required or not.

Reviewer: For auth fields this defaults to `true`, which should be documented.

`placeholder` | no | `string` | An example value that is not saved.
`default` | no | `string` | A default value that is saved the first time a Zap is created.

> `primary` | no | `boolean` | Use this field as part of the primary key for deduplication. You can set multiple fields as "primary", provided they are unique together. If no fields are set, Zapier will default to using the `id` field. `primary` only makes sense for `outputFields`; it will be ignored if set in `inputFields`. It only works in static `outputFields`; will not work in custom/dynamic `outputFields`. For more information, see [How deduplication works in Zapier](https://platform.zapier.com/build/deduplication).

Reviewer: Not relevant to auth fields.

> `dynamic` | no | [/RefResourceSchema](#refresourceschema) | A reference to a trigger that will power a dynamic dropdown.

Reviewer: This renders, but doesn't show choices, so should be dropped from this schema:

![](https://cdn.zappy.app/9c8ca5e68f5bb67e2f4da2758e6b6942.png)

> `search` | no | [/RefResourceSchema](#refresourceschema) | A reference to a search that will guide the user to add a search step to populate this field when creating a Zap.

Reviewer: Not relevant to auth fields.

`choices` | no | [/FieldChoicesSchema](#fieldchoicesschema) | An object of machine keys and human values to populate a static dropdown.
`list` | no | `boolean` | Acts differently when used in inputFields vs. when used in outputFields. In inputFields: Can a user provide multiples of this field? In outputFields: Does this field return an array of items of type `type`?
`children` | no | `array`[[/FieldSchema](#fieldschema)] | An array of child fields that define the structure of a sub-object for this field. Usually used for line items.
`dict` | no | `boolean` | Is this field a key/value input?
`computed` | no | `boolean` | Is this field automatically populated (and hidden from the user)? Note: Only OAuth and Session Auth support fields with this key.

> `altersDynamicFields` | no | `boolean` | Does the value of this field affect the definitions of other fields in the set?

Reviewer: Doesn't work in the auth dialog, so should be dropped until it does.

> `steadyState` | no | `boolean` | Prevents triggering on new output until all values for fields with this property remain unchanged for 2 polls. It can be used to, e.g., not trigger on a new contact until the contact has completed typing their name. NOTE that this only applies to the `outputFields` of polling triggers.

Reviewer: Not relevant to auth fields.

`inputFormat` | no | `string` | Useful when you expect the input to be part of a longer string. Put "{{input}}" in place of the user's input (IE: "https://{{input}}.yourdomain.com").

> `meta` | no | [/FieldMetaSchema](#fieldmetaschema) | Allows for additional metadata to be stored on the field. Supports simple key-values only (no sub-objects or arrays).

Reviewer: Not relevant to auth fields.

> `isSafe` | no | `boolean` | Indicates if this authentication field is safe (not secret).

Reviewer: I'm not sure about the name of this new property. Safe for what? Perhaps `isNoSecret` is more explicit? I know it violates some style guides because it is a negation, but I find it more clear.

Author: Good point, maybe `isConfidential` or `isSecret` convey the intent better without need for the negation?

Reviewer: @Natay I don't believe that is an option, as we currently consider all auth fields secret, so the new flag needs to opt fields out of that.

Author: Ahh that makes sense, I'll have it as `isNotSecret` and add a note then, thanks for the heads up!

Reviewer 2: @FokkeZB Fine to go with `isNoSecret`, but just wanted to point out that the description could do the work of explaining what `isSafe` means like what we currently have and still use it:

`Indicates if this authentication field is safe to e.g. be stored without encryption or displayed (not a secret).`

#### Properties

Key | Required | Type | Description
--- | -------- | ---- | -----------
`key` | **yes** | `string` | A unique machine readable key for this value (IE: "fname").
`label` | no | `string` | A human readable label for this value (IE: "First Name").
`helpText` | no | `string` | A human readable description of this value (IE: "The first part of a full name."). You can use Markdown.
`type` | no | `string` in (`'string'`, `'text'`, `'integer'`, `'number'`, `'boolean'`, `'datetime'`, `'file'`, `'password'`, `'copy'`, `'code'`) | The type of this value. Use `string` for basic text input, `text` for a large, `<textarea>` style box, and `code` for a `<textarea>` with a fixed-width font. Field type of `file` will accept either a file object or a string. If a URL is provided in the string, Zapier will automatically make a GET for that file. Otherwise, a .txt file will be generated.
`required` | no | `boolean` | If this value is required or not.
`placeholder` | no | `string` | An example value that is not saved.
`default` | no | `string` | A default value that is saved the first time a Zap is created.

> `primary` | no | `boolean` | Use this field as part of the primary key for deduplication. You can set multiple fields as "primary", provided they are unique together. If no fields are set, Zapier will default to using the `id` field. `primary` only makes sense for `outputFields`; it will be ignored if set in `inputFields`. It only works in static `outputFields`; will not work in custom/dynamic `outputFields`. For more information, see [How deduplication works in Zapier](https://platform.zapier.com/build/deduplication).

Reviewer: Not relevant for input fields and should does be dropped

`dynamic` | no | [/RefResourceSchema](#refresourceschema) | A reference to a trigger that will power a dynamic dropdown.
`search` | no | [/RefResourceSchema](#refresourceschema) | A reference to a search that will guide the user to add a search step to populate this field when creating a Zap.
`choices` | no | [/FieldChoicesSchema](#fieldchoicesschema) | An object of machine keys and human values to populate a static dropdown.
`list` | no | `boolean` | Acts differently when used in inputFields vs. when used in outputFields. In inputFields: Can a user provide multiples of this field? In outputFields: Does this field return an array of items of type `type`?
`children` | no | `array`[[/FieldSchema](#fieldschema)] | An array of child fields that define the structure of a sub-object for this field. Usually used for line items.
`dict` | no | `boolean` | Is this field a key/value input?

> `computed` | no | `boolean` | Is this field automatically populated (and hidden from the user)? Note: Only OAuth and Session Auth support fields with this key.

Reviewer: Not relevant for input fields and should does be dropped

`altersDynamicFields` | no | `boolean` | Does the value of this field affect the definitions of other fields in the set?

> `steadyState` | no | `boolean` | Prevents triggering on new output until all values for fields with this property remain unchanged for 2 polls. It can be used to, e.g., not trigger on a new contact until the contact has completed typing their name. NOTE that this only applies to the `outputFields` of polling triggers.

Reviewer: Not relevant for input fields and should does be dropped

-----

## /OutputFieldSchema

Field schema specialized for output fields.

#### Details

* **Type** - `object`
* [**Source Code**](https://github.com/zapier/zapier-platform/blob/zapier-platform-schema@16.2.0/packages/schema/lib/schemas/OutputFieldSchema.js)

#### Properties

Key | Required | Type | Description
--- | -------- | ---- | -----------
`key` | **yes** | `string` | A unique machine readable key for this value (IE: "fname").
`label` | no | `string` | A human readable label for this value (IE: "First Name").

> `helpText` | no | `string` | A human readable description of this value (IE: "The first part of a full name."). You can use Markdown.

Reviewer: Not relevant for output fields and should does be dropped.

> `type` | no | `string` in (`'string'`, `'text'`, `'integer'`, `'number'`, `'boolean'`, `'datetime'`, `'file'`, `'password'`, `'copy'`, `'code'`) | The type of this value. Use `string` for basic text input, `text` for a large, `<textarea>` style box, and `code` for a `<textarea>` with a fixed-width font. Field type of `file` will accept either a file object or a string. If a URL is provided in the string, Zapier will automatically make a GET for that file. Otherwise, a .txt file will be generated.

Reviewer: We should drop the types that make no sense for output fields, which at least are: `text` (as for output it's just about the value type, not the type of input field), `integer` (same), `copy` and `code`.

Not sue about `datetime` and `password`, but little harm in letting them be. And we'll need `file` for unfurl.

`required` | no | `boolean` | If this value is required or not.
`placeholder` | no | `string` | An example value that is not saved.

> `default` | no | `string` | A default value that is saved the first time a Zap is created.

Reviewer: Not relevant for input fields and should does be dropped.

> `default` | no | `string` | A default value that is saved the first time a Zap is created.

Reviewer: Not relevant for output fields and should does be dropped.

`primary` | no | `boolean` | Use this field as part of the primary key for deduplication. You can set multiple fields as "primary", provided they are unique together. If no fields are set, Zapier will default to using the `id` field. `primary` only makes sense for `outputFields`; it will be ignored if set in `inputFields`. It only works in static `outputFields`; will not work in custom/dynamic `outputFields`. For more information, see [How deduplication works in Zapier](https://platform.zapier.com/build/deduplication).
`dynamic` | no | [/RefResourceSchema](#refresourceschema) | A reference to a trigger that will power a dynamic dropdown.
`search` | no | [/RefResourceSchema](#refresourceschema) | A reference to a search that will guide the user to add a search step to populate this field when creating a Zap.

> `choices` | no | [/FieldChoicesSchema](#fieldchoicesschema) | An object of machine keys and human values to populate a static dropdown.

Reviewer: Not relevant for output fields and should does be dropped.

`list` | no | `boolean` | Acts differently when used in inputFields vs. when used in outputFields. In inputFields: Can a user provide multiples of this field? In outputFields: Does this field return an array of items of type `type`?
`children` | no | `array`[[/FieldSchema](#fieldschema)] | An array of child fields that define the structure of a sub-object for this field. Usually used for line items.
`dict` | no | `boolean` | Is this field a key/value input?
`computed` | no | `boolean` | Is this field automatically populated (and hidden from the user)? Note: Only OAuth and Session Auth support fields with this key.

> `altersDynamicFields` | no | `boolean` | Does the value of this field affect the definitions of other fields in the set?

Reviewer: Not relevant for output fields and should does be dropped.

`steadyState` | no | `boolean` | Prevents triggering on new output until all values for fields with this property remain unchanged for 2 polls. It can be used to, e.g., not trigger on a new contact until the contact has completed typing their name. NOTE that this only applies to the `outputFields` of polling triggers.
`inputFormat` | no | `string` | Useful when you expect the input to be part of a longer string. Put "{{input}}" in place of the user's input (IE: "https://{{input}}.yourdomain.com").

> `meta` | no | [/FieldMetaSchema](#fieldmetaschema) | Allows for additional metadata to be stored on the field. Supports simple key-values only (no sub-objects or arrays).

Reviewer: Not relevant for output fields and should does be dropped.
